From 3ca58a30c0cce790e06a418027af4ed3936f9f56 Mon Sep 17 00:00:00 2001
From: Rico Ganahl <rico.ganahl@bytesatwork.ch>
Date: Thu, 15 Jun 2023 11:05:07 +0200
Subject: [PATCH] spi: fsl_fspi: implement set speed

SPI is running at root clk speed which is too high in some cases.
Implement set speed in driver model ops.

Signed-off-by: Rico Ganahl <rico.ganahl@bytesatwork.ch>
---
 drivers/spi/fsl_fspi.c | 22 ++++++++++++++++++----
 1 file changed, 18 insertions(+), 4 deletions(-)

diff --git a/drivers/spi/fsl_fspi.c b/drivers/spi/fsl_fspi.c
index d6b272064c..350b25beea 100644
--- a/drivers/spi/fsl_fspi.c
+++ b/drivers/spi/fsl_fspi.c
@@ -157,6 +157,7 @@ struct fsl_fspi_priv {
 	u32 num_chipselect;
 	u32 read_dummy;
 	struct fsl_fspi_regs *regs;
+	struct clk clk;
 };
 
 struct fspi_cmd_func_pair {
@@ -1215,16 +1216,15 @@ static int fsl_fspi_probe(struct udevice *bus)
 
 	if (CONFIG_IS_ENABLED(CLK)) {
 		/* Assigned clock already set clock */
-		struct clk fspi_clk;
 		int ret;
 
-		ret = clk_get_by_name(bus, "fspi", &fspi_clk);
+		ret = clk_get_by_name(bus, "fspi", &priv->clk);
 		if (ret < 0) {
 			printf("Can't get fspi clk: %d\n", ret);
 			return ret;
 		}
 
-		ret = clk_enable(&fspi_clk);
+		ret = clk_enable(&priv->clk);
 		if (ret < 0) {
 			printf("Can't enable fspi clk: %d\n", ret);
 			return ret;
@@ -1391,7 +1391,21 @@ static int fsl_fspi_release_bus(struct udevice *dev)
 
 static int fsl_fspi_set_speed(struct udevice *bus, uint speed)
 {
-	/* Nothing to do */
+	if (CONFIG_IS_ENABLED(CLK)) {
+		struct fsl_fspi_priv *priv = dev_get_priv(bus);
+		int ret;
+
+		clk_disable(&priv->clk);
+
+		ret = clk_set_rate(&priv->clk, speed);
+		if (ret < 0)
+			return ret;
+
+		ret = clk_enable(&priv->clk);
+		if (ret)
+			return ret;
+	}
+
 	return 0;
 }
 
-- 
2.30.2

